rust/sdk-processor/src/steps/common/parquet_buffer_step.rs

```diff
@@ -0,0 +1,387 @@
+{
+    pub fn new(poll_interval: Duration, buffer_uploader: U, buffer_max_size: usize) -> Self {
+        Self {
+            internal_buffers: HashMap::new(),
+            poll_interval,
+            buffer_uploader,
+            buffer_max_size,
+        }
+    }
+
+    fn append_to_buffer(
+        buffer: &mut ParquetBuffer,
+        parquet_data: ParquetTypeStructs,
+    ) -> Result<(), ProcessorError> {
+        buffer.buffer_size_bytes += parquet_data.calculate_size();
+        buffer.buffer.append(parquet_data)
+    }
+
+    /// Handles the addition of `parquet_data` to the buffer for a specified `ParquetTypeEnum`.
+    ///
+    /// We check the size of the buffer + the size of the incoming data before appending it.
+    /// If the sum of the two exceeds the maximum limit size, it uploads the buffer content to GCS to avoid
+    /// spliting the batch data, allowing for more efficient and simpler version tracking.
+    async fn handle_buffer_append(
+        &mut self,
+        parquet_type: ParquetTypeEnum,
+        parquet_data: ParquetTypeStructs,
+        cur_batch_metadata: &TransactionMetadata,
+        upload_metadata_map: &mut HashMap<ParquetTypeEnum, TransactionMetadata>,
+    ) -> Result<(), ProcessorError> {
+        let mut file_uploaded = false;
+
+        // Get or initialize the buffer for the specific ParquetTypeEnum
+        let buffer = self
+            .internal_buffers
+            .entry(parquet_type)
+            .or_insert_with(|| {
+                debug!(
+                    "Initializing buffer for ParquetTypeEnum: {:?}",
+                    parquet_type,
+                );
+                ParquetBuffer::new(&parquet_type)
+            });
+
+        let curr_batch_size_bytes = parquet_data.calculate_size();
+
+        debug!(
+            "Current batch size for {:?}: {} bytes, buffer size before append: {} bytes",
+            parquet_type, curr_batch_size_bytes, buffer.buffer_size_bytes,
+        );
+
+        // If the current buffer size + new batch exceeds max size, upload the buffer
+        if buffer.buffer_size_bytes + curr_batch_size_bytes > self.buffer_max_size {
+            println!(
+                "Buffer size {} + batch size {} exceeds max size {}. Uploading buffer for {:?}.",
+                buffer.buffer_size_bytes, curr_batch_size_bytes, self.buffer_max_size, parquet_type
+            );
+
+            // Take the current buffer to upload and reset the buffer in place
+            let struct_buffer = std::mem::replace(
+                &mut buffer.buffer,
+                ParquetTypeStructs::default_for_type(&parquet_type),
+            );
+            self.buffer_uploader.handle_buffer(struct_buffer).await?;
+            file_uploaded = true;
+
+            // update this metadata before insert
+            upload_metadata_map
+                .insert(parquet_type, buffer.current_batch_metadata.clone().unwrap());
+            buffer.buffer_size_bytes = 0;
```

Comment:
I thin we also need to reset buffer.current_batch_metadata to None here?

Reply:
Good call!